### ISSUE TITLE
Mapping/Object Specific Whitespace Include Option

### ISSUE THREAD

**Issue by jamesb6626** [nvim-treesitter/nvim-treesitter-textobjects#282]
**Is your feature request related to a problem? Please describe.**
The select option `include_surrounding_whitespace` works great, but it applies to all objects and keymaps the same, which doesn't work well for inner objects in some languages. For example, in python, changing an inner function with `cif` with whitespace enabled drags the next function to the cursor, while disabling whitespace drops the nice feature for outer objects (same issue for classes).
Unfortunately, the opposite is needed for something like lua, where the inner function end is marked.

**Describe the solution you'd like**
For now, it would be great to be able to set `include_surrounding_whitespace` as a per-object option.
Maybe per mapping, but then what about `:TSTextobjectSelect`? Or per-filetype?
(maybe set object default and override for maps or filetypes)

**Describe alternatives you've considered**
At some point something more advanced might be needed, to try to figure out what the inner object should do.
Perhaps somehow, in the future, the plugin could figure out whether whitespace is needed for an inner object.
Maybe when the whitespace brings the selection to the beginning of another function/class it should be ignored?
(but would that work for other languages and situations?)
Otherwise, maybe per-filetype settings (e.g. lua vs python inner functions), and sensible per-language defaults would work.
But for now, being able to separately override the behaviour for different objects would be a nice step, and avoids breaking configs.

**Additional context**
Example: inner function (`vif`, `dif`, `cif`) on the second line of the following two files with `include_surrounding_whitespace` true/false.

Lua:
```
function()
  print("hello world")

end
```

Python:
```
def a():
    pass


def b():
    pass
```

**Comment by jamesb6626** [comment]
Just had a thought. What if, with the current setting enabled, the inner objects include whitespace, but only where that whitespace lies within the outer object selected without whitespace. The outer object includes the end marker (`end` for lua, `}` for C), but not the start of the next function in python. If something bad happens (are there cases with no outer? could outer not include inner?), fall back to no whitespace.

This would probably make inner selections more expensive (needs to calculate two objects) and less robust (might fail if either of the inner and outer selection does something strange), but hopefully gets the correct match.

Probably needs to leave the whitespace alone on the first line of the inner match (indent), first newline of preceding whitespace between outer and inner if it exists (newline after python def line is significant), and the final newline + indent of the final line (if exists) in the proceeding whitespace between the outer and inner (newline and indent for the end marker).
Basically, maybe try excluding one newline followed by indent in the whitespace that lies between the outer and inner matches at each end (first match at the start, last match at the end). Not sure whether that will work.

For languages where newlines are insignificant (C, lua) it probably makes sense to allow the user to optionally include all the inner whitespace (then `dif`, `cif` brings end brace to start brace, or `end` to `function()`, on the same line). But maybe that is something for the future.

**Comment by jamesb6626** [comment]
I'll probably close this for now, and reopen in a bit as a request specifically about inner objects whitespace.
I spotted #156 which will deal with the original request, assuming it allows `include_surrounding_whitespace` per mapping/object.
Also, per-filetype or more advanced user configured behaviour is maybe best achieved through #275, with user mappings/settings in the user's `ftplugin/*` or similar, assuming there is some way to use the setup config by default and override where needed.
